setup: Convert setup script to python

lsgunth opened the pull request:
Convert the setup script to python and add the ability to start
with loop disks or disks specified on the command line.

Martin: you can probably make use of this code in your work as well.

lsgunth committed:
setup: Convert setup script to python

Convert the setup script to python and add the ability to start
with loop disks or disks specified on the command line.

iomartin commented on md.py:
```diff
@@ -0,0 +1,91 @@
+        if ram_disks is not None:
+            if disks:
+                raise MDInvalidArgumentError("Must not specify both disks and ram_disks")
+            subprocess.check_call(["modprobe", "brd"])
```
This should be `modprobe brd rd_nr=count(ram_disks)` so that we create the right number of ram disks

iomartin replied:
Maybe use `rd_size=<bytes>` as well so that the disks aren't so tiny (the default is 4K)?

lsgunth replied:
The count won't work as some tests expect to grow more disks. Also the module may already be loaded so these parameters might be ignored... But I can add a "default" rd_size.

iomartin commented on md.py:
```diff
@@ -0,0 +1,91 @@
+            mdadm_args += ["--write-journal", journal]
+        if size is not None:
+            mdadm_args += ["--size", str(size)]
+
```
nit: two empty lines

iomartin commented on md.py:
```diff
@@ -0,0 +1,91 @@
+            mdadm_args += ["--size", str(size)]
+
+
+        print(mdadm_args + disks)
```
Is this a debug print that was left behind?

iomartin commented on setup:
```diff
@@ -1,72 +1,119 @@
+               chunk_size=args.chunk_size,
+               assume_clean=args.assume_clean,
+               force=args.force,
+               run=args.run or args.zero_first,
```
args.run and args.zero_first are only used here, can't they just be aliases to the same argument?

lsgunth replied:
"run" is a better name which matches mdadm but I've left zero_first for backwards compatibility with some of the other scripts.

iomartin replied:
But can't it be `parser.add_argument("--zero-first", "--run")`?

lsgunth replied:
The environement parser thingy only takes the dest argument as where to check the environment. So I want to check both environment variables.

iomartin commented on setup:
```diff
@@ -1,72 +1,119 @@
+    inst = md.MDInstance()
+    inst.setup(level=args.level,
+               ram_disks=args.disks,
+               loop_disks = args.loop_disks,
```
nit: extra space around =

iomartin commented on md.py:
```diff
@@ -0,0 +1,91 @@
+
+    def setup(self, level=5, disks=None, ram_disks=None, loop_disks=None,
+              size=None, chunk_size=64 << 10, assume_clean=True, force=True,
+              run=False, policy="resync", journal=False, quiet=False,
```
Did you mean `journal=None`?

This is later used as

```python
if journal is not None:
    mdadm_args += ["--write-journal", journal]
```

iomartin commented on md.py:
```diff
@@ -0,0 +1,91 @@
+                      "--chunk", str(chunk_size),
+                      "--raid-devices", str(len(disks))]
+
+        if policy == "bitmap":
```
How about the other policies?

iomartin commented on setup:
```diff
@@ -1,72 +1,119 @@
+
+    args = parser.parse_args()
+
+    if args.devs:
```
Isn't all of that guaranteed by the `parser.add_mutually_exclusive_group()`?

lsgunth replied:
No... because of the Environment variable parsing the defaults can bypass that.

iomartin commented on md.py:
```diff
@@ -0,0 +1,91 @@
+
+        if ram_disks is not None:
+            if disks:
+                raise MDInvalidArgumentError("Must not specify both disks and ram_disks")
```
Don't we need to check that ram_disks is also None?

lsgunth replied:
What?

iomartin replied:
Ops, I meant loop_disks

lsgunth replied:
disks gets set in loop disk so it doesn't need to explicitly also check ram_disks

dlsloan replied:
But shouldn't it validate the args before calling modprobe? seems weird to get part-way through init and then get an invalid argument exception

lsgunth replied:
Honestly, don't care if someone inadvertantly modprobes brd with a poorly constructed set of arguments.

iomartin commented on setup:
```diff
@@ -1,72 +1,119 @@
+                        help="cache size")
+    parser.add_argument("--journal", help="journal type")
+    parser.add_argument("--size", "--mdsize", dest="mdsize",
+                        type=suffix_parse, help="array size")
```
This should say that if --size is used with --loop-devices, then --size is the size of the loop devices

lsgunth replied:
Well it's actually the amount of bytes taken by each disk in both situations; so I've reworded it.

lsgunth committed:
test3: Use common md infrastructor and move md access code to md.py

Cleanup test3 to use the new python3 based setup script and
adjust things to work with loop devices.

lsgunth committed:
test_all: Add test with loop devices

Testing with loop devices uncovers some unusual stuff, so add a test
for that.

iomartin commented on md.py:
```diff
@@ -0,0 +1,204 @@
+            help = super()._get_help_string(action)
+            if action.dest != 'help':
+                help += ' [env: {}]'.format(action.dest.upper())
+                return help
```
The `return help` needs to be outside of the `if`

lsgunth committed:
test_all: Ensure test3 -g is run in default and skip second test2 run

Ensure test3 -g is run in default and skip second test2 run.

lsgunth committed:
test3: Catch keyboard interrupt

No need for a stack trace on ctrl-c

dlsloan commented on md.py:
```diff
@@ -0,0 +1,204 @@
+    def to_bool(self, x):
+        if isinstance(x, bool):
+            return x
+        if isinstance(x, str):
```
Can we add 1 for true too, that's fairly common too

iomartin commented on md.py:
```diff
@@ -5,6 +5,8 @@
 import pathlib
 import re
 import subprocess
+import sys
```
In the commit message:
s/then immediately stopping the arary./then immediately creating the array

lsgunth replied:
Actually hmmm. I'm investigating the possibility that the need for a delay here has been caused by a bug in a recent patch. I'm also failing an mdadm test all of a sudden.

iomartin reviewed (approved):
Just one small thing on a commit message but otherwise looks good to me

lsgunth commented:
Dropped the patch that was actually caused by a kernel regression so this can be merged.
